### PR TITLE
feat(testing): add flag for testing using compiled artifacts

### LINKED ISSUE #6253: Setup `build` as a target dependency for `test` and use compiled assets in tests
## Description

With the new version of Nx we can now define `build` for `dependencies` as a target that needs to be run prior to the original `build` command can be run using the `targetDependencies` in `nx.json`. It would be awesome to be able to use the compiled assets when running jests tests, rather than using jest-preset-angular with the `*.ts` transformer on a per spec file.

Essentially we would be able to put this in our `nx.json`
```
"targetDependencies": {
        "build": [
            {
                "target": "build",
                "projects": "dependencies"
            }
        ],
        "test": [
           {
               "target": "build",
               "projects": "dependencies"
            }
        ]
    }
```

and then when a spec file imports from TS such as `@scope/lib1` instead of pointing that alias to `/libs/lib1/src/index.ts` we would point it to `/dist/libs/lib1`, this is similar to how the Nrwl builder will work `--with-deps`

## Motivation

Should greatly speed up test runs if TypeScript files do not need to be compiled on the fly and takes advantage of the buildable deps Nx already provides.

## Suggested Implementation

`@nrwl/jest` would need to look at the `nx.json` in the same way that the builders do and determine if there are `targetDependencies` and then when the tests are run, the `jest.config.js` that is used will need to be modified such that we can override the TS paths. Not exactly how this should work since I know jest-preset-angular is used to do the transforming with ts-jest underneath, maybe we can pass in a temporary tsconfig file to those libraries so that we can merge the root `tsconfig.base.json` with the paths that were build.

`ts-jest` will look at the `moduleNameMapper` (https://kulshekhar.github.io/ts-jest/docs/getting-started/paths-mapping/) for path matching, `jest` supports passing in` moduleNameMapper` (althought the typings expect a `string` in `@types/jest` for `runCLI` so not sure how that would work) which should allow us to override where sources are mapped too (I am assuming that those would override paths in the tsconfig but I am not 100% sure on that)

### PR DESCRIPTION
**Pull request by yharaskrik**
## Current Behavior

When testing using @nrwl/jest right now ts-jest needs to compile all of the dependencies to be able to run the tests.

## Expected Behavior

Will be able to setup targetDependencies

```
"test": [
      {
        "target": "build",
        "projects": "dependencies"
      }
    ]
```

such that when running tests each of the libraries dependencies are built (or pulled from cache) and then those artifacts are used for testing instead of compiling on the fly using ts-jest. This allows us to leverage buildable libraries at test time rather than just at compile time.

## Related Issue(s)

Fixes #6253

### CODE REVIEW

**Comment by yharaskrik** [comment]
This also definitely still needs some tests, this is just a working version I tried in a local repo

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 1]
```diff
@@ -6,12 +6,86 @@ import { makeEmptyAggregatedTestResult, addResult } from '@jest/test-result';
...
+
+  tsConfig.extends = `${sourcePath}/tsconfig.json`;
+  // Will override the includes array from
+  // "include": ["**/*.test.ts", "**/*.spec.ts", "**/*.d.ts"]
+  // to:
+  // "include": ["../libs/lib1/**/*.test.ts", "../libs/lib1/**/*.spec.ts", "../libs/lib1/**/*.d.ts"]
+  tsConfig.include = tsConfig.include.map(
+    (includePath) => `${sourcePath}/${includePath}`
+  );
+  tsConfig.compilerOptions.paths = dependencies
```
[the remark this answers is not in the document]
Comment (yharaskrik) [reply packages/jest/src/executors/jest/jest.impl.ts:72]: Nvm, seems like the `calculateProjectDependencies` function does it recursively so it's all good!

**Comment by yharaskrik** [comment]
I have some updates locally that will fix the paths not being constructed correctly as well as correctly extending the jest.config.js rather than overwriting it so that settings from before are preserved. Just need to figure out how to build the path mappings for multiple entry points and across different output formats

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 2]
```diff
@@ -2,23 +2,109 @@ import 'dotenv/config';
...
+      const node = dep.node as ProjectGraphProjectNode;
+
+      const rootPath = node.data.root.split('/');
+      rootPath.shift();
+
+      const distPackageJson = readJsonFile(
+        path.join(context.cwd, 'dist', node.data.root, 'package.json')
+      );
+
+      const indexFile = distPackageJson['main'] ?? distPackageJson['esm2015'];
```
Comment (yharaskrik) [inline packages/jest/src/executors/jest/jest.impl.ts:68]: Wasn't sure what the best way is to find the main index.js in the repo, I am assuming node libs have `main` and angular packaged libs will have either main or default to the esm2015

**Comment by yharaskrik** [comment]
Having an issue with unit tests right now and not positive what to do to solve it

I am reading the project graph now (so that I can find the deps) but get this:

```
[readCachedProjectGraph] ERROR: No cached ProjectGraph is available.

    If you are leveraging \`readCachedProjectGraph()\` directly then you will need to refactor your usage to first ensure that
    the ProjectGraph is created by calling \`await createProjectGraphAsync()\` somewhere before attempting to read the data.

    If you encounter this error as part of running standard \`nx\` commands then please open an issue on https://github.com/nrwl/nx
```

But if I call createProjectGraphAsync before doing anything in the jest.impl.ts then I get this:

```
ENOENT: no such file or directory, open '/home/circleci/repo/tmp/unit/nx.json'

      32 |   options?: JsonReadOptions
      33 | ): T {
    > 34 |   const content = readFileSync(path, 'utf-8');
         |                               ^
      35 |   if (options) {
      36 |     options.endsWithNewline = content.charCodeAt(content.length - 1) === 10;
      37 |   }

      at readJsonFile (../tao/src/utils/fileutils.ts:34:31)
      at readNxJson (../workspace/src/core/file-utils.ts:206:30)
      at ../workspace/src/core/project-graph/project-graph.ts:51:28
      at ../../node_modules/tslib/tslib.js:117:75
      at __awaiter (../../node_modules/tslib/tslib.js:113:16)
      at createProjectGraphAsync (../workspace/src/core/project-graph/project-graph.ts:45:34)
```

I am assuming that is because the test workspace isn't setup the way a normal workspace would be?

**Comment by yharaskrik** [comment]
Looks like I fixed the testing issue. Everything is passing now. Would love some feedback on this from someone.

**Review by barbados-clemens — COMMENTED** [review]
Hey @yharaskrik thanks for the contribution and sorry for taking this long to get back to you. 

Being pretty new to this I am going to defer to @Cammisuli judgement on this feature. 

One thing that would be good to have is an e2e test making sure using `testFromSource: false` works.

**Comment by yharaskrik** [comment]
> Hey @yharaskrik thanks for the contribution and sorry for taking this long to get back to you. 
> 
> Being pretty new to this I am going to defer to @Cammisuli judgement on this feature. 
> 
> One thing that would be good to have is an e2e test making sure using `testFromSource: false` works.

I'll take a look at writing on e2e tests! I also realized I could probably use tsquery rather than appending to the end of the jest.config would probably be a better way to do it.

**Comment by yharaskrik** [comment]
Tested this PR against this repo which creates a few different libs across multiple generators and then import other libs into them. We then setup a targetDependency between test -> build so that when we test we build libs and the built artifacts are used in the tests

https://github.com/yharaskrik/nx-all-lib-type-buildable-test

**Comment by yharaskrik** [comment]
E2E test added!

**Comment by yharaskrik** [comment]
E2E tests pass locally but fail on CI... Would love some help tracking down the reason if possible! Thanks!

https://app.circleci.com/pipelines/github/nrwl/nx/13745/workflows/a766e3d4-93bc-4509-bab7-30e375add443/jobs/127781

**Comment by yharaskrik** [comment]
Nvm fixed it :)

File: packages/jest/src/executors/jest/jest.impl.ts [hunk 3]
```diff
@@ -2,23 +2,138 @@ import 'dotenv/config';
...
 ): Promise<{ success: boolean }> {
   const config = await jestConfigParser(options, context);
 
-  const { results } = await runCLI(config, [options.jestConfig]);
+  /*
+   * If config.config exists that means we are running tests using artifacts
+   * `testFromSource: false` otherwise default to the projects jest config
+   */
+  const { results } = await runCLI(config, [
+    config.config ?? options.jestConfig,
```
Comment (yharaskrik) [inline packages/jest/src/executors/jest/jest.impl.ts:135]: this is the one thing I am a little unsure on. I have it working for a single jest run but I didn't implement it for batch testing. Does that need to be done? Does does batch testing even work right now with nx, I have never seen anything referring to it.

**Comment by barbados-clemens** [comment]
going to close for now and we'll re-evaluate in the future and the benefits didn't seem to be there initially (speed wise)

**Comment by yharaskrik** [comment]
Agreed.

**Pull request closed without merging**